### PR TITLE
[KernelGen][Nvidia] Add empty operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `empty`. Tensor factory returning uninitialized memory, used for deferred init or output buffers.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_empty.py --level core` (NVIDIA CI)

### empty

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float16 | [1073741824] | 0.004096 | 0.005120 | 0.800 |
| float16 | [64, 64] | 0.004096 | 0.004096 | 1.000 |
| float16 | [4096, 4096] | 0.004096 | 0.004096 | 1.000 |
| float16 | [64, 512, 512] | 0.004096 | 0.004096 | 1.000 |
| float16 | [1024, 1024, 1024] | 0.004096 | 0.004096 | 1.000 |
| float32 | [1073741824] | 0.004096 | 0.004096 | 1.000 |
| float32 | [64, 64] | 0.004096 | 0.004096 | 1.000 |
| float32 | [4096, 4096] | 0.004096 | 0.004096 | 1.000 |
| float32 | [64, 512, 512] | 0.004096 | 0.004096 | 1.000 |
| float32 | [1024, 1024, 1024] | 0.004096 | 0.004096 | 1.000 |
| bfloat16 | [1073741824] | 0.004096 | 0.004096 | 1.000 |
| bfloat16 | [64, 64] | 0.004096 | 0.004096 | 1.000 |
| bfloat16 | [4096, 4096] | 0.004096 | 0.004096 | 1.000 |
| bfloat16 | [64, 512, 512] | 0.004096 | 0.004096 | 1.000 |
| bfloat16 | [1024, 1024, 1024] | 0.004096 | 0.004096 | 1.000 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| empty | **0.987** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (CI) | PASS (15 cases) | 0.987 | Primary |
| Tianshu | PASS | 0.999 | — |
| Muxi | PASS | 0.973 | — |
| Ascend | PASS | — | — |
| Hygon | PASS | 1.011 | — |

## Files Changed
- `src/flag_gems/ops/empty.py`: Triton kernel implementation
- `tests/test_empty.py`: Accuracy test
- `benchmark/test_empty.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Tensor, stage: alpha 5.1)

### CODE REVIEW

File: conf/operators.yaml [hunk 1]
```diff
@@ -6880,10 +6880,12 @@ ops:
...
       - alpha: '5.1'
   - id: empty
-    description: Tensor factory operation (empty).
+    description: Tensor factory returning uninitialized memory, used for deferred init or output buffers.
     for:
       - empty
     labels:
+      - aten
+      - KernelGen
       - skip_precision_check
```
Comment (Dongxu-H) [inline conf/operators.yaml]: this line could be deleted

File: tests/test_empty.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    with flag_gems.use_gems():
+        res_out = torch.empty(*shape, dtype=dtype, device=flag_gems.device)
+
+    ref_out = torch.zeros(*shape, dtype=dtype, device=expected_dev)
+    ref_out = utils.to_reference(ref_out, True)
+    utils.gems_assert_close(res_out, ref_out, dtype)
+
+
+@pytest.mark.empty
+def test_empty_default_dtype():
```
Comment (Dongxu-H) [inline tests/test_empty.py:26]: pls add comments to describe why this test function is needed

**Review by XDYuanzhuLee — COMMENTED** [review]
Fixed:
- Added comment to `test_empty_default_dtype` explaining it verifies proper dtype inference when only shape and device are given
- Reverted `operators.yaml` description to the simpler original `Tensor factory operation (empty).`

**Review by XDYuanzhuLee — COMMENTED** [review]
Resolved merge conflict in operators.yaml with upstream/master. Kept the empty operator entry with alpha: 5.3 stage (matching upstream convention). Dropped stale entries that upstream removed.

File: conf/operators.yaml (hunk 1 above)
Comment (XDYuanzhuLee) [reply conf/operators.yaml]: The line in question is the skip_precision_check label. It is necessary because empty() returns uninitialized memory — precision comparisons against a reference output are not applicable for this operator.

File: tests/test_empty.py (hunk 2 above)
Comment (XDYuanzhuLee) [reply tests/test_empty.py:26]: Fixed. Added comment explaining that test_empty_default_dtype verifies proper dtype inference when only shape and device are given.

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator 贡献审查报告

**Operator**: `empty`
**PR**: #3693

### 📊 变更统计
- 新增行数: 130
- 删除行数: 0  
- 变更文件: 6

### ✅ 检查清单

| 检查项 | 状态 |
|--------|------|
| 实现文件 (`src/flag_gems/ops/empty.py`) | ✅ 存在 |
| 测试文件 (`tests/test_empty.py`) | ✅ 存在 |
| Benchmark 文件 | ✅ 存在 |
| 配置注册 (`operators.yaml`) | ✅ 存在 |
| 模块注册 (`__init__.py`) | ✅ 存在 |
| KernelGen 标注 | ✅ 存在 |
| Triton JIT 装饰器 | ✅ 存在 |
| Metax Backend | ⚠️ 未实现 |

### 📝 审查意见

**代码结构**: 实现文件结构正确

**测试覆盖**: 有测试文件

**性能测试**: 有 benchmark

---

*此审查由自动化脚本根据 FlagGems Operator 贡献规范生成。*
*如有问题请联系维护者进行详细审查。*

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `empty`  
**PR**: #3693

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 130 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/empty.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ✅ Present: `tests/test_empty.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_empty.py` |
| 10 | Benchmark implementation correctness | ⚠️ Benchmark may be incomplete |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ⚠️ Needs domain expert review |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ✅ Edge case tests present |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ✅ Docstrings present |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 12 |
| ⚠️ Warning | 9 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Caeruleann — DISMISSED** [review]

**Comment by Dongxu-H** [comment]
## 代码审查结果 - PR #3693

### ✅ 通过的检查

1. **CI 工作流状态**：`python-op` 和 `code-style` 通过
2. **合并冲突**：无冲突 (MERGEABLE)
3. **配置一致性**：`conf/operators.yaml` 排序正确（`empty` 在 `elu_backward` 之后，`ilshift__` 之前）
4. **Kernel 注册**：`__init__.py` 导入和注册位置正确（按字母顺序：`embedding_dense_backward` → `empty` → `eq`)
5. **Logger 命名**：正确使用 `logging.getLogger(__name__)`（主文件夹位置）
6. **测试覆盖**：
   - 多种 shape 和 dtype 测试 ✓
   - 默认 dtype 推断测试 ✓

7. **代码实现**：
   - 正确处理 dtype 默认值推断 ✓
   - 正确处理 device 默认值 ✓
   - 使用 `torch.zeros` + kernel 作为 empty 实现（与测试参考一致）

---

**总体评价**：代码质量良好，实现正确，测试覆盖完整。可以合并。

**Review by Dongxu-H — DISMISSED** [review]

**Comment by XDYuanzhuLee** [comment]
感谢审查！所有检查均已通过，conf/operators.yaml 中 empty 条目的排序问题已在最新提交中修正。可以合并。

**Comment by XDYuanzhuLee** [comment]
感谢审查！所有检查均已通过，conf/operators.yaml 中 empty 条目的排序问题已在最新提交中修正。可以合并。

**Review by Dongxu-H — APPROVED** [review]

**Review by Caeruleann — APPROVED** [review]
